src/StreamJsonRpc/Reflection/TargetMethod.cs

```diff
@@ -15,36 +15,38 @@ namespace StreamJsonRpc
 
     internal sealed class TargetMethod
     {
-        private readonly HashSet<string> errorMessages = new HashSet<string>(StringComparer.Ordinal);
         private readonly JsonRpcMessage request;
         private readonly object target;
         private readonly MethodInfo method;
         private readonly object[] parameters;
 
+        /// <summary>
+        /// A collection of error messages. May be null until the first message is added.
+        /// </summary>
+        private HashSet<string> errorMessages;
+
         internal TargetMethod(
             JsonRpcMessage request,
             JsonSerializer jsonSerializer,
-            IEnumerable<MethodSignatureAndTarget> candidateMethodTargets)
+            List<MethodSignatureAndTarget> candidateMethodTargets)
         {
             Requires.NotNull(request, nameof(request));
             Requires.NotNull(jsonSerializer, nameof(jsonSerializer));
             Requires.NotNull(candidateMethodTargets, nameof(candidateMethodTargets));
 
             this.request = request;
 
-            var targetMethods = new Dictionary<MethodSignatureAndTarget, object[]>();
             foreach (var method in candidateMethodTargets)
             {
-                this.TryAddMethod(request, targetMethods, method, jsonSerializer);
-            }
-
-            KeyValuePair<MethodSignatureAndTarget, object[]> methodWithParameters = targetMethods.FirstOrDefault();
-            if (methodWithParameters.Key.Signature != null)
-            {
-                this.target = methodWithParameters.Key.Target;
-                this.method = methodWithParameters.Key.Signature.MethodInfo;
-                this.parameters = methodWithParameters.Value;
-                this.AcceptsCancellationToken = methodWithParameters.Key.Signature.HasCancellationTokenParameter;
+                object[] args = this.TryGetParameters(request, method.Signature, jsonSerializer, request.Method);
+                if (this.method == null && args != null)
```

Comment:
Can you rename the local method variable to candidateMethod so that it is not confused by "this.method" :)

Reply:
Done